src/ota_image_libs/v1/resource_table/utils.py

```diff
@@ -121,35 +133,80 @@ def _thread_local_dctx(self) -> zstandard.ZstdDecompressor:
             self._thread_local.dctx = dctx
             return dctx
 
+    def _prepare_one_bundle(
+        self, _bundle_f: Path, _bundle_entry: ResourceTableManifest
+    ):
+        # try to re-use already prepared bundle file resource
+        if (
+            _bundle_f.is_file()
+            and file_sha256(_bundle_f).digest() == _bundle_entry.digest
+        ):
+            return
+
+        _bundle_save_tmp = self._download_dir / tmp_fname(
+            str(_bundle_entry.resource_id)
+        )
+        yield from self._prepare_resource(_bundle_entry, _bundle_save_tmp)
+        # unconditionally override the previous file
+        os.replace(_bundle_save_tmp, _bundle_f)
+
     def _prepare_bundled_resource(
         self, entry: ResourceTableManifest, save_dst: Path
     ) -> Generator[ResourceDownloadInfo]:
-        assert isinstance(entry.filter_applied, BundleFilter)
-        # NOTE: prevent the same bundle being prepared again and again
-        bundle_rsid = entry.filter_applied.list_resource_id()
+        filter_cfg = entry.filter_applied
+        assert isinstance(filter_cfg, BundleFilter)
+
+        # NOTE: only do one query for each bundle
+        bundle_rsid = filter_cfg.list_resource_id()
         with self._bundle_process_lock:
-            if bundle_rsid not in self._bundle:
-                bundle_entry = self._orm_pool.orm_select_entry(resource_id=bundle_rsid)
-                logger.debug(f"Requesting bundle({bundle_rsid=}): {bundle_entry}")
-                _bundle_save_tmp = self._download_dir / tmp_fname(str(bundle_rsid))
-                bundle_save_dst = self._download_dir / bundle_entry.digest.hex()
-
-                yield from self._prepare_resource(bundle_entry, _bundle_save_tmp)
-                os.replace(_bundle_save_tmp, bundle_save_dst)
-                self._bundle[bundle_rsid] = bundle_save_dst
-            bundle_fpath = self._bundle[bundle_rsid]
-
-        # NOTE: keep the bundle for later use, but if recreate from bundle failed, we delete the bundle
-        #       to let otaclient do the re-downloading.
+            if not (_bundle_info := self._bundle_per_locks.get(bundle_rsid)):
+                _bundle_entry = self._orm_pool.orm_select_entry(resource_id=bundle_rsid)
```

Comment:
`orm_select_entry` might return `None`. How about to check None right after this code? For example,
```
                if _bundle_entry is None:
                    raise BundledRecreateFailed(
                        f"bundle resource {bundle_rsid} not found in resource table"
                    )
```

Reply:
Good catch! Although we must not allow this issue occurs, this issue means the OTA image has some problem :fearful: .